chain/chain/src/chain.rs

```diff
@@ -788,74 +791,365 @@ impl Chain {
         self.store.get_outgoing_receipts_for_shard(prev_block_hash, shard_id, last_height_included)
     }
 
-    pub fn state_request(
+    pub fn get_state_for_shard(
         &mut self,
+        _me: &Option<AccountId>,
         shard_id: ShardId,
-        hash: CryptoHash,
-    ) -> Result<(ChunkExtra, Vec<u8>, ReceiptResponse, Vec<ReceiptResponse>), Error> {
-        let prev_hash = self.get_block_header(&hash)?.inner.prev_hash;
-
-        let prev_block = self.get_block(&prev_hash)?;
+        sync_hash: CryptoHash,
+    ) -> Result<
+        (
+            ShardChunk,
+            MerklePath,
+            Vec<u8>,
+            Vec<SignedTransaction>,
+            Vec<ReceiptProofResponse>,
+            Vec<Vec<RootProof>>,
+        ),
+        Error,
+    > {
+        // Consistency rules:
+        // 1. Everything prefixed with `sync_` indicates new epoch, for which we are syncing.
+        // 1a. `sync_prev` means the last of the prev epoch.
+        // 2. Empty prefix means the height where chunk was applied last time in the prev epoch.
+        //    Let's call it `current`.
+        // 2a. `prev_` means we're working with height before current.
+        // 3. In inner loops we use all prefixes with no relation to the context described above.
+        let sync_block = self.get_block(&sync_hash)?;
+        let sync_block_header = sync_block.header.clone();
+        if shard_id as usize >= sync_block.chunks.len() {
+            return Err(ErrorKind::Other("Invalid request: ShardId out of bounds".into()).into());
+        }
 
-        if shard_id as usize >= prev_block.chunks.len() {
+        // The chunk was applied at height `chunk_header.height_included`.
+        // Getting the `current` state.
+        let sync_prev_block = self.get_block(&sync_block_header.inner.prev_hash)?;
+        if shard_id as usize >= sync_prev_block.chunks.len() {
             return Err(ErrorKind::Other("Invalid request: ShardId out of bounds".into()).into());
         }
+        // Chunk header here is the same chunk header of `current` height.
```

Comment:
"... is the same chunk header as at the `current` height"

Reply:
++